Add optional save button to Attachment element

squelart opened the pull request:
#### 7dbe55762352e84e4b449ae9733b1efdafc50c6d
<pre>
Add optional save button to Attachment element
<a href="https://bugs.webkit.org/show_bug.cgi?id=252420">https://bugs.webkit.org/show_bug.cgi?id=252420</a>
rdar://105250269

Reviewed by Aditya Keerthi and Tim Nguyen.

* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::attachmentSaveButtonIdentifier):
(WebCore::saveAttr):
(WebCore::createContainedElement):
(WebCore::HTMLAttachmentElement::ensureModernShadowTree):
(WebCore::HTMLAttachmentElement::updateSaveButton):
(WebCore::HTMLAttachmentElement::parseAttribute):
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(button#attachment-save-button):

Canonical link: <a href="https://commits.webkit.org/260640@main">https://commits.webkit.org/260640@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9287f820761aac26b9066148d2be99c87e585b11

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108906 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/18011 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41749 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/452 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/118224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112803 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/19467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9295 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101156 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114676 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/19467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97824 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/42728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/19467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29465 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/84468 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10785 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30814 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11539 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/16929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50410 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7351 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13129 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -147,27 +158,56 @@ void HTMLAttachmentElement::ensureModernShadowTree(ShadowRoot& root)
         m_elementWithSubtitle->setInnerText(WTFMove(subtitle));
-    container->appendChild(*m_elementWithSubtitle);
+    m_elementContainer->appendChild(*m_elementWithSubtitle);
+
```
fwiw, instead of multiple appendChild calls, you can do `m_elementContainer->append(X, Y, Z)` which appends everything at once.

squelart replied:
I don't see that function.
There's `ContainerNode::append(FixedVector<NodeOrString>&& vector)`, it doesn't seem used a lot, and it would do a lot of work to construct this `FixedVector`, then convert it to a small tree of nodes, and that would append that tree to the container. If that's what you were thinking of, are you sure that's better?

nt1m replied:
Yeah I was thinking of this function, though I realize this is C++ so it's not as fancy as when you call it from JS 😅 

Feel free to ignore this.

squelart replied:
Thanks. But you made me notice the repetitions; I'll see if I can refactor these similar-looking blocks...

squelart replied:
I've refactored this whole section, please have a look.

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -147,27 +158,56 @@ void HTMLAttachmentElement::ensureModernShadowTree(ShadowRoot& root)
+    saveDispatch.append("dispatchEvent(new CustomEvent('"_s);
+    saveDispatch.append(eventTypeName);
+    saveDispatch.append("', { bubbles: true, composed: true }))"_s);
+    m_elementWithSave->setAttributeWithoutSynchronization(HTMLNames::onclickAttr, saveDispatch.toAtomString());
```
This seems error prone and prone to security vulnerabilities as well.

Can create a new event listener class that has an `handleEvent` function that calls `dispatchEvent` from C++? 

`PDFDocument.cpp` might have some good examples.

pxlcoder replied:
`SpinButtonElement` is another good example.

squelart replied:
Updated with custom listener, thanks for the help.

nt1m commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -62,3 +62,15 @@ div#attachment-subtitle {
+    grid-column: 3;
+    align-self: center;
+    /* SF Symbol square.and.arrow.down */
+    background-image: url('data:image/svg+xml,<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="17.334" height="21.5723"><path d="M3.06641 21.084L14.2676 21.084C16.3086 21.084 17.334 20.0684 17.334 18.0566L17.334 8.31055C17.334 6.29883 16.3086 5.2832 14.2676 5.2832L11.543 5.2832L11.543 6.85547L14.2383 6.85547C15.2051 6.85547 15.7617 7.38281 15 ...
```
You don't really need the xlink namespace:

```suggestion
    background-image: url('data:image/svg+xml,<svg version="1.1" xmlns="http://www.w3.org/2000/svg" width="17.334" height="21.5723"><path d="M3.06641 21.084L14.2676 21.084C16.3086 21.084 17.334 20.0684 17.334 18.0566L17.334 8.31055C17.334 6.29883 16.3086 5.2832 14.2676 5.2832L11.543 5.2832L11.543 6.85547L14.2383 6.85547C15.2051 6.85547 15.7617 7.38281 15.7617 8.39844L15.7617 17.9688C15.7617 18.9844 15.2051 19.5117 14.2383 19.5117L3.08594 19.5117C2.10938 19.5117 1.57227 18.9844 1.57227 17.9688L1.57227 8.39844C1.57227 7.38281 2.10938 6.85547 3.08594 6.85547L5.79102 6.85547L5.79102 5.2832L3.06641 5.2832C1.02539 5.2832 0 6.29883 0 8.31055L0 18.0566C0 20.0684 1.02539 21.084 3.06641 21.084ZM8.66211 14.6582C8.86719 14.6582 9.0332 14.5996 9.22852 14.4043L12.5293 11.2109C12.6758 11.0645 12.7637 10.9082 12.7637 10.7031C12.7637 10.3027 12.4512 10.0195 12.0508 10.0195C11.8555 10.0195 11.6602 10.0977 11.5234 10.2539L10.0391 11.8262L9.38477 12.5195L9.44336 11.0547L9.44336 0.761719C9.44336 0.351562 9.08203 0 8.66211 0C8.24219 0 7.89062 0.351562 7.89062 0.761719L7.89062 11.0547L7.94922 12.5195L7.28516 11.8262L5.81055 10.2539C5.67383 10.0977 5.45898 10.0195 5.27344 10.0195C4.86328 10.0195 4.57031 10.3027 4.57031 10.7031C4.57031 10.9082 4.64844 11.0645 4.79492 11.2109L8.0957 14.4043C8.30078 14.5996 8.4668 14.6582 8.66211 14.6582Z" stroke="black"/></svg>');
```

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -146,28 +175,74 @@ void HTMLAttachmentElement::ensureModernShadowTree(ShadowRoot& root)
+    if (!m_elementWithSave)
+        m_elementWithSave = createShadowDivElement(attachmentSaveIdentifier(), String { });
+
+    m_elementWithSave->addEventListener(eventNames().clickEvent, AttachmentSaveEventListener::create(*this, eventTypeName), { });
```
So if we change multiple times the "save" attribute, do we end up having multiple event listeners?

squelart replied:
Oh, whoops, good catch! 😅

nt1m replied:
You can just have a single type of `AttachmentSaveEventListener` that reads off the save attribute fwiw, instead of passing in the save attribute to the event listener. That should remove the need to remove/re-add the listener.

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -121,6 +124,32 @@ static const AtomString& attachmentSubtitleIdentifier()
+{
+    element->setIdAttribute(id);
+    if (!innerText.isEmpty())
+        element->setInnerText(WTFMove(innerText));
```
nit: `setTextContent()` (`.textContent =` in JS) is faster if you're not relying on innerText specific behaviors.

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -146,28 +175,72 @@ void HTMLAttachmentElement::ensureModernShadowTree(ShadowRoot& root)
+    if (!m_elementWithSave) {
+        m_elementWithSave = createShadowDivElement(attachmentSaveIdentifier(), String { });
+        m_elementWithSave->addEventListener(eventNames().clickEvent, AttachmentSaveEventListener::create(*this), { });
+    }
```
```suggestion
    if (!m_elementContainer)
        return;
        
    if (m_elementWithSave) {
        if (eventTypeName.isEmpty()) {
            m_elementContainer->removeChild(*m_elementWithSave);
            m_elementWithSave = nullptr;
        }
        return;
    }

    m_elementWithSave = createShadowDivElement(attachmentSaveIdentifier(), String { });
    m_elementWithSave->addEventListener(eventNames().clickEvent, AttachmentSaveEventListener::create(*this), { });
```

squelart replied:
Your suggested code doesn't correctly handle the case where `m_elementWithSave` is null and the `eventTypeName` is empty: It would go to line 237 and always create the save button, when it shouldn't (because of the empty `eventTypeName`). We'd need to add `if (!eventTypeName.isEmpty())` around lines 237-238, which goes back to the double-if that I think you were trying to avoid.

nt1m replied:
Ah I guess your code is fine then. Btw, we should probably check against a null string rather than an empty string if possible.

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -146,28 +175,72 @@ void HTMLAttachmentElement::ensureModernShadowTree(ShadowRoot& root)
+    }
+
+    if (!m_elementWithSave) {
+        m_elementWithSave = createShadowDivElement(attachmentSaveIdentifier(), String { });
```
For accessibility purposes, should this be a `<button>`? It requires a bit more styling, but it's the correct semantics at least.

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.h:
```diff
@@ -119,9 +122,11 @@ class HTMLAttachmentElement final : public HTMLElement {
     RefPtr<HTMLElement> m_elementWithAction;
     RefPtr<HTMLElement> m_elementWithTitle;
     RefPtr<HTMLElement> m_elementWithSubtitle;
+    RefPtr<HTMLElement> m_elementWithSave;
```
I should have probably brought this up before, but I preferred if we named these differently:

```suggestion
    RefPtr<HTMLAttachmentElement> m_attachmentIcon;
    RefPtr<HTMLElement> m_containerElement;
    RefPtr<HTMLElement> m_actionTextElement;
    RefPtr<HTMLElement> m_titleElement;
    RefPtr<HTMLElement> m_subtitleElement;
    RefPtr<HTMLElement> m_saveButton;
```

squelart replied:
Good idea, thanks.
Though I'll keep `m_innerLegacyAttachment` for now (because it currently does more than just display an icon) but I'll add a mention in the FIXME about changing the name when we'll get rid of the inner-attachment trick.

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -146,28 +175,72 @@ void HTMLAttachmentElement::ensureModernShadowTree(ShadowRoot& root)
+            init.bubbles = true;
+            init.cancelable = true;
+            init.composed = true;
+            m_attachment->dispatchEvent(CustomEvent::create(m_attachment->attributeWithoutSynchronization(saveQualifiedName()), init, event.isTrusted() ? Event::IsTrusted::Yes : Event::IsTrusted::No));
```
Feels kinda weird that this just emits a custom event that has what's inside the save attribute, but I assume you coordinated with other teams using `<attachment>` and ended up with this solution.

squelart replied:
I didn't coordinate with the Mail team yet, but they're going to start using it soon, so I'll hopefully get feedback about that!
Since this element is inside a user agent shadow DOM, it's not reachable by their code, so they can't attach a listener directly to it; and I can't just pass a standard `click` event because there would be no way to distinguish it from a click on the whole attachment brick -- but please let me know if I'm incorrect.
And I wanted to give our client a way to potentially distinguish Save-clicks from different attachments (just in case), that's why I'm using the `save` attribute value to name the custom event.

What else would you recommend, that would feel less weird?

nt1m replied:
It purely depends on what the client needs. From a web developer POV, I would find a single event name easier to handle, with the detail of what's being saved in the event detail/data, simply because I would be able to watch them from a single event listener.


Although if there are no specific needs, a `<a download>` tag pointing to a blob might just be fine as well.

nt1m replied:
If this were JS, I'd do: `new CustomEvent("save", { detail: [...] });`

squelart replied:
Thanks for the suggestion. I'll iterate with the client devs, see if they'd like this or something else...

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -121,6 +124,32 @@ static const AtomString& attachmentSubtitleIdentifier()
+    return QualifiedName { nullAtom(), "save"_s, nullAtom() };
+}
+
+void HTMLAttachmentElement::addShadowElement(RefPtr<HTMLElement> element, const AtomString& id, String innerText)
```
Can we just combine `createShadowDivElement` and `addShadowElement` ?

`addShadowElement` is an odd function to call on its own.

squelart replied:
I needed `addShadowElement` for adding the inner attachment, because it's constructed in a special way that `createShadowDivElement` cannot do.

nt1m replied:
My point is that it feels odd even when building inner attachment, I'd rather have a bit more duplication here.

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -146,28 +175,72 @@ void HTMLAttachmentElement::ensureModernShadowTree(ShadowRoot& root)
+    void handleEvent(ScriptExecutionContext&, Event& event) final
+    {
+        if (event.type() == eventNames().clickEvent) {
+            event.preventDefault();
+            event.stopPropagation();
+            event.stopImmediatePropagation();
```
I assume this is to prevent the default attachment click behavior, but are all three of these needed?

squelart replied:
I don't know, I blindly copied it from another event listener (probably [`MediaControlsContextMenuEventListener::handleEvent`](https://searchfox.org/wubkat/rev/dacbd7c36056647c1867f34236ade6815002e1f1/Source/WebCore/Modules/mediacontrols/MediaControlsHost.cpp#455-457)). 🤷

nt1m replied:
Usually these are trying to prevent the default click event listener from triggering, so they should only be needed if the attachment element itself has a click listener, and you don't want it to run it at the same time as the save button listener.

squelart replied:
That's the case here: If the user clicks on the save button, it gets converted into a custom event out of the top-level attachment, and I don't want the initial click to invoke click event listeners on that attachment (or above).

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.h:
```diff
@@ -93,6 +93,9 @@ class HTMLAttachmentElement final : public HTMLElement {
 
     void didAddUserAgentShadowRoot(ShadowRoot&) final;
     void ensureModernShadowTree(ShadowRoot&);
+    void updateSave(const AtomString& eventTypeName);
```
Can we name this `updateSaveButton` ?

nt1m reviewed (commented):
Looks good, mostly nits.

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -121,24 +124,46 @@ static const AtomString& attachmentSubtitleIdentifier()
+    RefPtr<ElementType> element = ElementType::create(document());
+    element->setIdAttribute(id);
+    if (!innerText.isEmpty())
+        element->setTextContent(WTFMove(innerText));
```
nit: rename innerText to textContent

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -121,24 +124,46 @@ static const AtomString& attachmentSubtitleIdentifier()
+
+static QualifiedName saveQualifiedName()
+{
+    return QualifiedName { nullAtom(), "save"_s, nullAtom() };
```
I wonder if this should be added to `HTMLAttributeNames.in`? We do that for `subtitleAttr` which is specific to `<attachment>`.

If not, I think we should still name this method as `saveAttr`.

cc @whsieh @hortont424

nt1m replied:
I'd prefer to reserve `HTMLAttributeNames.in` to commonly web platform items, I don't think we need to optimize for `<attachment>`. `subtitleAttr` should be scoped to HTMLAttachmentElement.cpp as well fwiw

whsieh replied:
I think both approaches sound reasonable — but either way we choose, I do think it'd be good to make `subtitleAttr` consistent with this new attribute.

squelart replied:
I'll change `saveQualifiedName` to `saveAttr` here.
For `subtitleAttr`: I've filed rdar://105697448 (subtitleAttr should be scoped to HTMLAttachmentElement.cpp)

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -121,24 +124,46 @@ static const AtomString& attachmentSubtitleIdentifier()
     return identifier;
 }
 
+static const AtomString& attachmentSaveIdentifier()
```
`attachmentSaveButtonIdentifier`?

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -121,24 +124,46 @@ static const AtomString& attachmentSubtitleIdentifier()
 
+static const AtomString& attachmentSaveIdentifier()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("attachment-save"_s);
```
`attachment-save-button`?

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -121,24 +124,46 @@ static const AtomString& attachmentSubtitleIdentifier()
+}
+
+template <typename ElementType>
+RefPtr<ElementType> HTMLAttachmentElement::createShadowElement(const AtomString& id, String textContent)
```
Not sure this is the best name, given that it does more than just create the element, it also adds to the container. 

It might be better to just inline the calls to `appendChild`. Especially because right now this method doesn't null check or assert `m_containerElement`. The method is easy to misuse in its current state.

nt1m replied:
I think the factoring is good, we can add an assert if necessary.

squelart replied:
I'll pass the container by ref, rename the function to `createAndAddElementTo`, and then it won't depend on any member data so it can be made static. And I'll return a `Ref<>`.

nt1m replied:
I'm not sure I like this new name, the calls don't read like a sentence:
`createAndAddElementTo<HTMLDivElement>(attachmentActionIdentifier(), attachmentActionForDisplay(), *m_containerElement);`

Create and add element to HTML div element, attachment action identifier, action for display.

squelart replied:
We need ObjC named parameters!
`createElementAndAppendTo`?
`createElementWithIdAndOptionalTextAndAppendTo`? (Getting a bit long...)

squelart replied:
`createContainedElement<Div>(container, id, text)`?

squelart committed:
Add optional save button to Attachment element
https://bugs.webkit.org/show_bug.cgi?id=252420
rdar://105250269

Reviewed by Aditya Keerthi and Tim Nguyen.

* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::attachmentSaveButtonIdentifier):
(WebCore::saveAttr):
(WebCore::createContainedElement):
(WebCore::HTMLAttachmentElement::ensureModernShadowTree):
(WebCore::HTMLAttachmentElement::updateSaveButton):
(WebCore::HTMLAttachmentElement::parseAttribute):
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(button#attachment-save-button):

Canonical link: https://commits.webkit.org/260640@main
